CondFormats/HcalObjects/interface/HcalConvertedPedestalWidthsGPU.h

```diff
@@ -0,0 +1,42 @@
+#ifndef CondFormats_HcalObjects_interface_HcalConvertedPedestalWidthsGPU_h
+#define CondFormats_HcalObjects_interface_HcalConvertedPedestalWidthsGPU_h
+
+#include "CondFormats/HcalObjects/interface/HcalPedestals.h"
+#include "CondFormats/HcalObjects/interface/HcalPedestalWidths.h"
+#include "CondFormats/HcalObjects/interface/HcalQIEData.h"
+#include "CondFormats/HcalObjects/interface/HcalQIETypes.h"
+
+#ifndef __CUDACC__
+#include "HeterogeneousCore/CUDAUtilities/interface/HostAllocator.h"
+#include "HeterogeneousCore/CUDACore/interface/ESProduct.h"
+#endif
+
+class HcalConvertedPedestalWidthsGPU {
+public:
+  struct Product {
+    ~Product();
+    float* values;
+  };
+
+#ifndef __CUDACC__
+  // order matters!
+  HcalConvertedPedestalWidthsGPU(HcalPedestals const&,
+                                 HcalPedestalWidths const&,
+                                 HcalQIEData const&,
+                                 HcalQIETypes const&);
+
+  // will trigger deallocation of Product thru ~Product
+  ~HcalConvertedPedestalWidthsGPU() = default;
+
+  // get device pointers
+  Product const& getProduct(cudaStream_t) const;
```

Comment:
do you think that changing the definition of `Product` to
```c++
  struct Product {
    ~Product();
    edm::propagate_const<float*> values;
  };
```
would fix this potential issue ?

Reply:
I believe `edm::propagate_const` would solve issue. I suppose it would require declaring ~all methods of `propagate_const` as `constexpr` for it to work in the device code.

Reply:
Actually this changes compile fine:
```diff
diff --git a/CondFormats/HcalObjects/interface/HcalConvertedPedestalWidthsGPU.h b/CondFormats/HcalObjects/interface/HcalConvertedPedestalWidthsGPU.h
index e3adb59ca54e..1afad7ca3a66 100644
--- a/CondFormats/HcalObjects/interface/HcalConvertedPedestalWidthsGPU.h
+++ b/CondFormats/HcalObjects/interface/HcalConvertedPedestalWidthsGPU.h
@@ -5,6 +5,7 @@
 #include "CondFormats/HcalObjects/interface/HcalPedestalWidths.h"
 #include "CondFormats/HcalObjects/interface/HcalQIEData.h"
 #include "CondFormats/HcalObjects/interface/HcalQIETypes.h"
+#include "FWCore/Utilities/interface/propagate_const.h"
 
 #ifndef __CUDACC__
 #include "HeterogeneousCore/CUDAUtilities/interface/HostAllocator.h"
@@ -15,7 +16,7 @@ class HcalConvertedPedestalWidthsGPU {
 public:
   struct Product {
     ~Product();
-    float* values;
+    edm::propagate_const<float*> values;
   };
 
 #ifndef __CUDACC__
diff --git a/CondFormats/HcalObjects/src/HcalConvertedPedestalWidthsGPU.cc b/CondFormats/HcalObjects/src/HcalConvertedPedestalWidthsGPU.cc
index 1d7ca5b3394c..0af2b5eaade2 100644
--- a/CondFormats/HcalObjects/src/HcalConvertedPedestalWidthsGPU.cc
+++ b/CondFormats/HcalObjects/src/HcalConvertedPedestalWidthsGPU.cc
@@ -137,7 +137,7 @@ HcalConvertedPedestalWidthsGPU::HcalConvertedPedestalWidthsGPU(HcalPedestals con
 
 HcalConvertedPedestalWidthsGPU::Product::~Product() {
   // deallocation
-  cudaCheck(cudaFree(values));
+  cudaCheck(cudaFree(values.get()));
 }
 
 HcalConvertedPedestalWidthsGPU::Product const& HcalConvertedPedestalWidthsGPU::getProduct(
@@ -145,10 +145,12 @@ HcalConvertedPedestalWidthsGPU::Product const& HcalConvertedPedestalWidthsGPU::g
   auto const& product = product_.dataForCurrentDeviceAsync(
       cudaStream, [this](HcalConvertedPedestalWidthsGPU::Product& product, cudaStream_t cudaStream) {
         // malloc
-        cudaCheck(cudaMalloc((void**)&product.values, this->values_.size() * sizeof(float)));
+        float* values;
+        cudaCheck(cudaMalloc(&values, values_.size() * sizeof(float)));
+        product.values = values;
 
         // transfer
-        cudaCheck(cudaMemcpyAsync(product.values,
+        cudaCheck(cudaMemcpyAsync(product.values.get(),
                                   this->values_.data(),
                                   this->values_.size() * sizeof(float),
                                   cudaMemcpyHostToDevice,
```

Reply:
> Actually these changes compile fine:

Ah, right, probably `HcalConvertedPedestalWidthsGPU::Product` is not passed to device code as such, but the individual arrays themselves.

Reply:
@makortel can you confirm that the implementation that uses `edm::propagate_const_array` is now fine ?

Reply:
> can you confirm that the implementation that uses `edm::propagate_const_array` is now fine ?

Visually the use of `edm::propagate_const_array` looks good now, thanks.